{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Corona Vaccine Recombinant Expression Simulation\n",
    "***\n",
    "\n",
    "## Background\n",
    "\n",
    "The Covid19 associated pandemic is affecting societies and new vaccines are needed to provide lasting immunity. Traditionally, vaccination strategies based on protein subunits have been popular ([Jeyanathan et al., 2020](https://doi.org/10.1038/s41577-020-00434-6)), which are effectively produced by bacterial recombinant expression ([He et al., 2017](https://doi.org/10.1139/cjm-2016-0528), [Strizova et al.,2021](https://doi.org/10.1159/000514225)). You are member of a small biotech company, specialized in recombinant expression systems and your task is to develop a bacterial host to produce the highest possible amount of viral protein subunits. The expression of the viral protein is controlled by important sequences in the promoter and the final vaccine production rate depends on reaching a high biomass by cultivating in optimal temperatures. Alas, you have only a limited amount of money. You have to aquire the starting material equipment and each experiment costs resources.\n",
    "\n",
    "<figure>\n",
    "    <img src=\"../Figures/Jupyter/Jeyanathan_Covid19-Vaccines_20_Fig1.png\" width=\"500\">\n",
    "    <figcaption>The largest share of Corona vaccines use recombinantly expressed protein subunits (<a href='https://doi.org/10.1038/s41577-020-00434-6'>Jeyanathan et al., 2020</a>). </figcaption>\n",
    "</figure>\n",
    "\n",
    "In this project, selected steps of a biotechnological project for recombinant expressions are simulated. The experimental work in your biotech company is highly automated: you are setting the parameters of the experiments and focus on computational data analysis. The goal is to optimize the production rate of the protein subunits to be competitive compared to your peers. To achieve this goal, virtual experiments have to be performed to optimize growth conditions and promoter sequence for the production. Finally, a comparison will be performed how the GC-content of the promoter affects the promoter activity. The data analysis can be performed either separately with Excel, or with some guided coding steps within this script. \n",
    "\n",
    "You start with a budget of 10.000 EUR. The initial laboratory setup and each subsequent experiment is associated with an investment. To optimize your initial host-promoter combination towards more effective production, a number of 4-6 strains might be necessary. Initially, you decide on how much money to spend on the laboratory equipment, investing too little will result in a higher failure rate of experiments. Some steps are difficult to perform. The exact parameters for effective cloning are unknown and depend on various complex factors.\n",
    "\n",
    "|Initial Budget: 10.000 EUR|\n",
    "| --- |\n",
    "\n",
    "\n",
    "| Experiment | Cost in EUR |\n",
    "| --- | --- |\n",
    "| Equipment | 10-20% of budget |\n",
    "| Temperature growth | 100 |\n",
    "| Cloning | 200 |\n",
    "| Promoter Strength | 100 |\n",
    "| Production run | 500 |\n",
    "\n",
    "\n",
    "## Workflow\n",
    "\n",
    "**1 Set-up of simulation environment**\n",
    "\n",
    "**2 Lab setup**\n",
    " \n",
    " * *2.1 Choose your host organism*\n",
    " * *2.2 Choose Equipment investment*\n",
    " \n",
    "**3 Culture characterization**\n",
    " \n",
    " * *3.1 Experiment set-up*     \n",
    " * *3.2 Data analysis growth experiment*    \n",
    "\n",
    "**4 Promoter sequence selection**\n",
    " * *4.1 Promoter and expression experiments*\n",
    " * *4.2 Data analysis of promoter strength*\n",
    "\n",
    "**5 Evaluation by cross-group integration**\n",
    "\n",
    "## 1 Set-up of simulation environment\n",
    "Loading libraries and fixing visualization. No user input necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: silvio in /home/ulf/Documents/2109_BLS+SILVIO/silvio/src (from -r ../requirements.txt (line 1)) (0.2.6)\n",
      "Collecting biopython==1.79\n",
      "  Downloading biopython-1.79-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.7 MB)\n",
      "     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.7/2.7 MB 5.7 MB/s eta 0:00:00\n",
      "Requirement already satisfied: cobra in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 3)) (0.29.0)\n",
      "Collecting IPython==7.20\n",
      "  Using cached ipython-7.20.0-py3-none-any.whl (784 kB)\n",
      "Requirement already satisfied: joblib>=0.14 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 6)) (1.1.0)\n",
      "Requirement already satisfied: matplotlib in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 7)) (3.5.1)\n",
      "Collecting plotly\n",
      "  Downloading plotly-6.0.0-py3-none-any.whl (14.8 MB)\n",
      "     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.8/14.8 MB 5.8 MB/s eta 0:00:00\n",
      "Requirement already satisfied: numpy in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 9)) (1.23.5)\n",
      "Requirement already satisfied: openpyxl in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 10)) (3.0.9)\n",
      "Requirement already satisfied: pandas in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 11)) (2.2.2)\n",
      "Requirement already satisfied: recordclass in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 12)) (0.17.2)\n",
      "Requirement already satisfied: charset-normalizer~=2.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 13)) (2.0.0)\n",
      "Collecting scikit-learn==1.2.2\n",
      "  Downloading scikit_learn-1.2.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (9.6 MB)\n",
      "     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.6/9.6 MB 6.0 MB/s eta 0:00:00\n",
      "Requirement already satisfied: scipy in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 15)) (1.9.3)\n",
      "Requirement already satisfied: watermark in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from -r ../requirements.txt (line 18)) (2.3.0)\n",
      "Collecting bokeh~=3.1.1\n",
      "  Downloading bokeh-3.1.1-py3-none-any.whl (8.3 MB)\n",
      "     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.3/8.3 MB 5.5 MB/s eta 0:00:00\n",
      "Collecting biocircuits==0.1.11\n",
      "  Using cached biocircuits-0.1.11-py2.py3-none-any.whl (65 kB)\n",
      "Collecting wget\n",
      "  Using cached wget-3.2-py3-none-any.whl\n",
      "Requirement already satisfied: pygments in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (2.14.0)\n",
      "Requirement already satisfied: setuptools>=18.5 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (67.6.0)\n",
      "Requirement already satisfied: backcall in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (0.2.0)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (3.0.32)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (0.18.1)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (4.8.0)\n",
      "Requirement already satisfied: traitlets>=4.2 in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (5.9.0)\n",
      "Requirement already satisfied: decorator in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (5.1.1)\n",
      "Requirement already satisfied: pickleshare in /home/ulf/.local/lib/python3.9/site-packages (from IPython==7.20->-r ../requirements.txt (line 5)) (0.7.5)\n",
      "Collecting joblib>=0.14\n",
      "  Downloading joblib-1.4.2-py3-none-any.whl (301 kB)\n",
      "     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.8/301.8 kB 5.4 MB/s eta 0:00:00\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/ulf/.local/lib/python3.9/site-packages (from scikit-learn==1.2.2->-r ../requirements.txt (line 14)) (3.1.0)\n",
      "Requirement already satisfied: eqtk in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from biocircuits==0.1.11->-r ../requirements.txt (line 24)) (0.1.3)\n",
      "Requirement already satisfied: multiprocess in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from biocircuits==0.1.11->-r ../requirements.txt (line 24)) (0.70.14)\n",
      "Requirement already satisfied: numba in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from biocircuits==0.1.11->-r ../requirements.txt (line 24)) (0.56.4)\n",
      "Requirement already satisfied: tqdm in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from biocircuits==0.1.11->-r ../requirements.txt (line 24)) (4.65.0)\n",
      "Requirement already satisfied: colorcet in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from biocircuits==0.1.11->-r ../requirements.txt (line 24)) (3.0.1)\n",
      "Requirement already satisfied: jinja2<3.1.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from silvio->-r ../requirements.txt (line 1)) (3.0.3)\n",
      "Requirement already satisfied: python-libsbml~=5.19 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (5.19.0)\n",
      "Requirement already satisfied: httpx~=0.24 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (0.27.0)\n",
      "Requirement already satisfied: future in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (0.18.2)\n",
      "Requirement already satisfied: importlib-resources in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (5.12.0)\n",
      "Requirement already satisfied: rich>=8.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (12.2.0)\n",
      "Requirement already satisfied: optlang~=1.8 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (1.8.1)\n",
      "Requirement already satisfied: depinfo~=2.2 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (2.2.0)\n",
      "Requirement already satisfied: ruamel.yaml~=0.16 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (0.17.21)\n",
      "Requirement already satisfied: swiglpk in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (5.0.10)\n",
      "Requirement already satisfied: appdirs~=1.4 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (1.4.4)\n",
      "Requirement already satisfied: diskcache~=5.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (5.4.0)\n",
      "Requirement already satisfied: pydantic>=1.6 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from cobra->-r ../requirements.txt (line 3)) (1.9.0)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (9.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/ulf/.local/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (3.0.9)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (4.31.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/ulf/.local/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (21.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/ulf/.local/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (2.8.2)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (1.4.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from matplotlib->-r ../requirements.txt (line 7)) (0.11.0)\n",
      "Collecting narwhals>=1.15.1\n",
      "  Downloading narwhals-1.25.0-py3-none-any.whl (313 kB)\n",
      "     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 313.3/313.3 kB 5.7 MB/s eta 0:00:00\n",
      "Requirement already satisfied: et-xmlfile in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from openpyxl->-r ../requirements.txt (line 10)) (1.1.0)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages/tzdata-2024.1-py3.9.egg (from pandas->-r ../requirements.txt (line 11)) (2024.1)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from pandas->-r ../requirements.txt (line 11)) (2022.7.1)\n",
      "Requirement already satisfied: xyzservices>=2021.09.1 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from bokeh~=3.1.1->-r ../requirements.txt (line 23)) (2023.2.0)\n",
      "Requirement already satisfied: PyYAML>=3.10 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from bokeh~=3.1.1->-r ../requirements.txt (line 23)) (6.0)\n",
      "Requirement already satisfied: contourpy>=1 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from bokeh~=3.1.1->-r ../requirements.txt (line 23)) (1.0.7)\n",
      "Requirement already satisfied: tornado>=5.1 in /home/ulf/.local/lib/python3.9/site-packages (from bokeh~=3.1.1->-r ../requirements.txt (line 23)) (6.2)\n",
      "Requirement already satisfied: certifi in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from httpx~=0.24->cobra->-r ../requirements.txt (line 3)) (2022.12.7)\n",
      "Requirement already satisfied: sniffio in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from httpx~=0.24->cobra->-r ../requirements.txt (line 3)) (1.2.0)\n",
      "Requirement already satisfied: httpcore==1.* in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from httpx~=0.24->cobra->-r ../requirements.txt (line 3)) (1.0.5)\n",
      "Requirement already satisfied: idna in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from httpx~=0.24->cobra->-r ../requirements.txt (line 3)) (3.3)\n",
      "Requirement already satisfied: anyio in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from httpx~=0.24->cobra->-r ../requirements.txt (line 3)) (3.5.0)\n",
      "Requirement already satisfied: h11<0.15,>=0.13 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from httpcore==1.*->httpx~=0.24->cobra->-r ../requirements.txt (line 3)) (0.14.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.0 in /home/ulf/.local/lib/python3.9/site-packages (from jedi>=0.16->IPython==7.20->-r ../requirements.txt (line 5)) (0.8.3)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from jinja2<3.1.0->silvio->-r ../requirements.txt (line 1)) (2.1.2)\n",
      "Requirement already satisfied: sympy>=1.12.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from optlang~=1.8->cobra->-r ../requirements.txt (line 3)) (1.12)\n",
      "Requirement already satisfied: six>=1.9 in /home/ulf/.local/lib/python3.9/site-packages (from optlang~=1.8->cobra->-r ../requirements.txt (line 3)) (1.16.0)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/ulf/.local/lib/python3.9/site-packages (from pexpect>4.3->IPython==7.20->-r ../requirements.txt (line 5)) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /home/ulf/.local/lib/python3.9/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->IPython==7.20->-r ../requirements.txt (line 5)) (0.2.5)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/ulf/.local/lib/python3.9/site-packages (from pydantic>=1.6->cobra->-r ../requirements.txt (line 3)) (4.5.0)\n",
      "Requirement already satisfied: commonmark<0.10.0,>=0.9.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from rich>=8.0->cobra->-r ../requirements.txt (line 3)) (0.9.1)\n",
      "Requirement already satisfied: ruamel.yaml.clib>=0.2.6 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from ruamel.yaml~=0.16->cobra->-r ../requirements.txt (line 3)) (0.2.6)\n",
      "Requirement already satisfied: pyct>=0.4.4 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from colorcet->biocircuits==0.1.11->-r ../requirements.txt (line 24)) (0.5.0)\n",
      "Requirement already satisfied: zipp>=3.1.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from importlib-resources->cobra->-r ../requirements.txt (line 3)) (3.15.0)\n",
      "Requirement already satisfied: dill>=0.3.6 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from multiprocess->biocircuits==0.1.11->-r ../requirements.txt (line 24)) (0.3.6)\n",
      "Requirement already satisfied: llvmlite<0.40,>=0.39.0dev0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from numba->biocircuits==0.1.11->-r ../requirements.txt (line 24)) (0.39.1)\n",
      "Requirement already satisfied: param>=1.7.0 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from pyct>=0.4.4->colorcet->biocircuits==0.1.11->-r ../requirements.txt (line 24)) (1.13.0)\n",
      "Requirement already satisfied: mpmath>=0.19 in /home/ulf/anaconda3/envs/bls_py39/lib/python3.9/site-packages (from sympy>=1.12.0->optlang~=1.8->cobra->-r ../requirements.txt (line 3)) (1.2.1)\n",
      "Installing collected packages: wget, narwhals, joblib, biopython, scikit-learn, plotly, IPython, bokeh, biocircuits\n",
      "  Attempting uninstall: joblib\n",
      "    Found existing installation: joblib 1.1.0\n",
      "    Uninstalling joblib-1.1.0:\n",
      "      Successfully uninstalled joblib-1.1.0\n",
      "  Attempting uninstall: biopython\n",
      "    Found existing installation: biopython 1.83\n",
      "    Uninstalling biopython-1.83:\n",
      "      Successfully uninstalled biopython-1.83\n",
      "  Attempting uninstall: scikit-learn\n",
      "    Found existing installation: scikit-learn 1.0.2\n",
      "    Uninstalling scikit-learn-1.0.2:\n",
      "      Successfully uninstalled scikit-learn-1.0.2\n",
      "  Attempting uninstall: IPython\n",
      "    Found existing installation: ipython 8.12.0\n",
      "    Uninstalling ipython-8.12.0:\n",
      "      Successfully uninstalled ipython-8.12.0\n",
      "  Attempting uninstall: bokeh\n",
      "    Found existing installation: bokeh 2.4.3\n",
      "    Uninstalling bokeh-2.4.3:\n",
      "      Successfully uninstalled bokeh-2.4.3\n",
      "  Attempting uninstall: biocircuits\n",
      "    Found existing installation: biocircuits 0.1.8\n",
      "    Uninstalling biocircuits-0.1.8:\n",
      "      Successfully uninstalled biocircuits-0.1.8\n",
      "Successfully installed IPython-7.34.0 biocircuits-0.1.11 biopython-1.79 bokeh-3.1.1 joblib-1.4.2 narwhals-1.25.0 plotly-6.0.0 scikit-learn-1.2.2 wget-3.2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "ipykernel 6.17.0 requires ipython>=7.23.1, but you have ipython 7.20.0 which is incompatible.\n",
      "pydna 5.2.0 requires biopython>=1.80, but you have biopython 1.79 which is incompatible.\n",
      "panel 0.14.4 requires bokeh<2.5.0,>=2.4.0, but you have bokeh 3.1.1 which is incompatible.\n",
      "escher 1.7.3 requires ipywidgets<8,>=7.4.0, but you have ipywidgets 8.0.4 which is incompatible.\n",
      "escher 1.7.3 requires Jinja2<3,>=2.7.3, but you have jinja2 3.0.3 which is incompatible.\n",
      "escher 1.7.3 requires jsonschema<4,>=3.0.1, but you have jsonschema 4.17.3 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "# Adapting environment to GitHub or Google Colab\n",
    "# In GitHub, the full repository is available with all directories, and libraries can be installed with pip on requirements.txt\n",
    "# In Google Colab only the selected notebook is loaded and the requirements file must be downloaded from GitHub before installation.\n",
    "# \n",
    "\n",
    "# file system and path operations\n",
    "import os\n",
    "import sys\n",
    "if 'google.colab' in sys.modules:\n",
    "    IN_COLAB = True\n",
    "    # Download the requirements file\n",
    "    os.system('wget https://raw.githubusercontent.com/biolabsim/BioLabSim/refs/heads/master/requirements.txt')\n",
    "    os.system('pip install -r requirements.txt')\n",
    "else:\n",
    "    IN_COLAB = False\n",
    "    os.system('pip install -r ../requirements.txt')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "System ready\n"
     ]
    }
   ],
   "source": [
    "# Allow imports from the current directory.\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "pd.options.display.width = 100 # Make pandas output wider.\n",
    "\n",
    "# Release only. Disable warnings.\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore') # Comment this line to show warnings.\n",
    "\n",
    "# Release only. Disable the traceback when exceptions occur (like the *_or_abort methods)\n",
    "ipython = get_ipython()\n",
    "def hide_traceback(exc_tuple=None, filename=None, tb_offset=None,\n",
    "                   exception_only=False, running_compiled_code=False):\n",
    "    etype, value, tb = sys.exc_info()\n",
    "    return ipython._showtraceback(etype, value, ipython.InteractiveTB.get_exception_only(etype, value))\n",
    "ipython.showtraceback = hide_traceback  # Comment this line to show traceback on exceptions.\n",
    "\n",
    "from silvio.catalog.RecExpSim import RecExperiment, combine_data\n",
    "from silvio.extensions.records.gene.crafted_gene import CraftedGene\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# additional parameter for experiment\n",
    "bpar = 10000\n",
    "wpar = 0.005\n",
    "\n",
    "print('System ready')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 Lab setup\n",
    "\n",
    "In this stage, you decide on which host organism to use for your recombinant expression system and the investment to the laboratory equipment. You have the choice between two organisms for recombinant expression, namely *E. coli* (abbr. `ecol`) and *P. putida* (abbr. `pput`). A high investment in the laboratory equipment increases the probability of successfull experiments, \n",
    "\n",
    "The two most popular prokarytic organisms for biotechnology are *E. coli* and *P. putida*. *E. coli* is an intensely studied model organism, and allows easy manipulation and fast growth. *P. putida* is a soil bacterium with a high metabolic versatility, including xenobiotic degradation. It shows a very high robustness against extreme environmental conditions such as high temperature, extreme pH, or the presence of toxins or inhibiting solvents. \n",
    "\n",
    "**Resource cost:**\n",
    "* **Free** \n",
    "\n",
    "**Input:** \n",
    "* **`mySeed`: Number, e.g., Student-ID (integer)**\n",
    "* **`myInvest`: 1000-2000 Eur (10-20% of total budget, integer)**\n",
    "* **`myMutant`: 'ecol' or 'pput' (string)**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# User input is required in the following code lines:\n",
    "\n",
    "# Initialize the random seed with a number, e.g., student-ID\n",
    "Seed = None\n",
    "# Enter here the investment for the equipment, higher investment results in fewer experiment failures\n",
    "myInvest = None\n",
    "# To choose the host organism replace None in the 'Mutant'-command with the abbreviation:\n",
    "myMutant = None\n",
    "\n",
    "# SILVIO experiment initialization\n",
    "exp = RecExperiment( Seed, myInvest, bpar )\n",
    "host = exp.create_host(myMutant, name='origin')\n",
    "\n",
    "# host organism and remaining resources are displayed:\n",
    "exp.print_status()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3 Culture characterization\n",
    "\n",
    "### 3.1  Experiment set-up\n",
    "You have to identify the optimal growth temperature, the corresponding maximum growth rate and the maximum biomass of your strain by cultivating the cells at different temperatures. For each program start the optimal temperature and the maximum biomass is randomly initiated. The optimal temperature is sampled from the range of temperatures for **mesophilic microorganisms** within 20-40°C (see page 23, [Biotechnology: An illustrated primer](https://application.wiley-vch.de/books/sample/3527335153_c01.pdf)). Occasionally, a culture will not grow at all and therefore it might be helpful to measure temperature replicates. However, be aware that each cultivation costs resources.     \n",
    "\n",
    "The results from the temperature experiments are stored in a comma separated value (csv) file  in the `Data` subfolder as `Growth_Simulation.csv`. You can find the csv-file in the left navigation panel within the `biolabsim` folder. For quick inspection double click the csv-file, for downloading and subsequent analysis in Excel right click on the file name and choose 'Download'.        \n",
    "\n",
    "If you want to do another set of experiments afterwards, or if you want to repeat individual experiments, you should make sure that you change the ID of your set of experiments (experiments_ID), otherwise results already generated may be overwritten. By default the experiments_ID has the value `1`, as is shown in the code cell below. \n",
    "\n",
    "Example: `temperature = [22,24,26,28,30,32,34,36,38]`\n",
    "\n",
    "**Resource cost:**\n",
    "* **100 EUR for each Experiment**\n",
    "\n",
    "**Input:**\n",
    "* **`temperatures`: Temperature array (integer list)**\n",
    "* **`experiments_ID`: variable name (integer)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# User input is required in the following code lines:\n",
    "temperatures = None\n",
    "\n",
    "# SILVIO simulation\n",
    "growth_out = exp.simulate_growth('origin', temperatures, wpar)\n",
    "filepath = os.path.join('..','Data','Growth_Simulation.csv')\n",
    "growth_out.export_data(filepath)\n",
    "\n",
    "exp.print_status()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Data analysis growth experiment\n",
    "\n",
    "The data of the experiment was stored in a coma-separated-value (`csv`) file in the local adress. The data has to be analysed to extract optimal temperature, growth rate and maximum biomass. You have the choice to either analyse the data via a spreadsheet application on your local computer, e.g. Excel, or via a programming approach with Python.\n",
    "\n",
    "In Excel you have to import the csv-file to get the data into separate columns. You then apply a natural logarithm to the data and plot the value versus time. The plot allows to extract the temperature supporting fastest growth with highest slope and the time until which the increase of logarithmic biomass is linear, i.e. it displays exponential growth. Then you apply a linear regression on the linear section of the fastest logarithmic biomass increase to extract the growth rate as the regression slope. You determine the maximum biomass by averaging over a number of measurements on the plateau of the measured biomass (real values, not logarithm).\n",
    "\n",
    "For the Python approach, there are scambled solutions of lines of code which you have to organize correctly. Even without programming experience this should take an equivalent time compared to the Excel approach. Please record the time it takes for you to conduct the data analysis and proceed to the item `Promoter sequence selection`.\n",
    "\n",
    "#### 3.2.2 Python based growth analysis\n",
    "\n",
    "In the following, data analysis via Python can be performed. The procedure is separated in two steps: 1. visualization of the results, 2. extracting growth rate and maximum biomass for the optimal temperature. For each process the corresponding lines of code are provided in the cell, but you have to arrange them in right order.\n",
    "\n",
    "##### 3.2.2.1 Visual analysis of exponential growth\n",
    "\n",
    "The visual indication of exponential growth is a linear slope in the plot of the logarithm of biomass versus time. The following commands are involved (scrambled order):\n",
    " * assign variables for time and biomass data\n",
    " * loading of the csv-file into a numpy data array ([genfromtxt](https://stackoverflow.com/questions/3518778/how-do-i-read-csv-data-into-a-record-array-in-numpy))\n",
    " * plotting biomass versus time ([plt.scatter](https://matplotlib.org/2.0.2/users/pyplot_tutorial.html))\n",
    " * store natural logarithm of the biomass in new variable ([np.log](https://numpy.org/doc/stable/reference/generated/numpy.log.html))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insert the correct code sequence for plotting in this cell.\n",
    "# %load Snippets/rev_GrowthPlot.py \n",
    "Time, Biomass = my_data[:,0], my_data[:,1:]\n",
    "DataFile = os.path.join('..','Data','Growth_Simulation.csv')\n",
    "LnBiomass = np.log(Biomass)\n",
    "[plt.scatter(Time, X, label=Exp) for Exp,X in enumerate(LnBiomass.T)]\n",
    "plt.legend([r'{}:{}$^\\circ$C'.format(Idx, T) for Idx, T in enumerate(temperatures)], bbox_to_anchor=(1.05, 1), loc='upper left'); plt.xlabel('time, h'); plt.ylabel('ln(Biomass)')\n",
    "my_data = np.genfromtxt(DataFile, delimiter=',', skip_header=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 3.2.2.2 Determine maximum biomass and growth rate\n",
    "\n",
    "After having the graphical insight, select the optimal temperature with fastest growth using the experiment index number in Python. Be carefull, Python starts counting with zero! Thus, the first experiment has the index `0`. Also, extract from the half-logarithmic plot the time range of linear growth as integer number (e.g. `10`h). This value will be used to estimate the growth rate via linear regression. The process is described in the following scrambled lines:\n",
    " * extract the mean biomass after the linear slope as max biomass ([np.mean](https://numpy.org/doc/stable/reference/generated/numpy.mean.html))\n",
    " * identify the latest time (integer number) of linear slope\n",
    " * conduct linear regression in the linear region ([np.polyfit](https://numpy.org/doc/stable/reference/generated/numpy.polyfit.html))\n",
    " \n",
    "Like in the previous coding example you have to bring the code in right order.\n",
    "\n",
    "**Input:**\n",
    "* **1st None: Index of Optimum Temperature (integer)**\n",
    "* **2nd None: max time in h for linear slope (integer)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rearrange the correct code sequence for calculating the growth and biomass parameters here.\n",
    "# For None enter the corresponding values of experiment index with fastest growth and latest time of linear growth (integer number).\n",
    "# %load Snippets/snip_GrowthPars.py\n",
    "\n",
    "Idx_optT, Linear_optT = None, None\n",
    "MB = np.mean(Biomass[Linear_optT:,Idx_optT])\n",
    "print('max biomass: {:.0f}\\nmax growth rate: {:.2f}'.format(MB,GR[0]))\n",
    "GR = np.polyfit(Time[:Linear_optT],LnBiomass[:Linear_optT,Idx_optT],1)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4 Promoter sequence selection\n",
    "In bacteria, the initiation of transcription at promoters requires the sigma factor to bind the RNA polymerase core to form the holoenzyme. Sigma factors recognize and open the promoter DNA and perform the initial steps in RNA synthesis. Particularly important DNA recognition sites are the -10 box and -35 box positions. An important bacterial sigma factor is sigma70 and directs the bulk of transcription during active growth [Paget, 2015](https://doi.org/10.3390/biom5031245). Two sites in the promoter are particularly important to determine the transcription strength: -35 and -10 boxes. The boxes are conserved sequences with a defined number of nucleotides upstream of the transcription start site. A common nucleotide sequence for the -35 box is: `TTGACA`, and for the -10 box: `TATAAT`. \n",
    "\n",
    "In this simulation, the total length of the promoters must be 40 nt. The following template is a starting point for deriving promoters. Replace the **X** with nucleotides and identify the best sequence (hit: the six **X** at start and end represent -35 and -10 box, respectively):\n",
    "\n",
    " * GCCCA**XXXXXX**A**X**GC**XXX**C**X**CGT**XXX**GG**XXXXXX**TGCACG\n",
    "\n",
    "### 4.1 Promoter and expression experiments\n",
    "\n",
    "#### 4.1.1 Promoter choice and cloning\n",
    "To introduce the promoter into your organism, you have to conduct a cloning and design a suitable primer for each promoter. First create the primer that is complementary to your promoter sequences. The primers should always start at the first nucleotide of the promoter sequences. You have to identify the optimal primer length, which can vary from 15-30 nucleotides. Because the primer will still work if it deviates by 5 nucleotides, you can make broad jumps to test the right primer length.\n",
    "\n",
    "Calculate the melting temperature of the primer according to the formula [genelink manual](https://www.genelink.com/Literature/ps/R26-6400-MW.pdf):\n",
    " * Tm = 2(A+T) + 4(C+G)\n",
    "\n",
    "**Important note/warning:** \n",
    "The success of cloning is a probabilistic event, even with the right conditions, it might fail in a first trial but succeed in a second identical trial. Note the meaning of the following failure reports:\n",
    " * Cloning failed: \n",
    "   * promoter deviates from 40nt?\n",
    "   * primer fully complementary to promoter and starts with the first nucleotide?\n",
    "   * primer length incorrect?\n",
    "   * melting temperature incorrect?\n",
    "   * just try two more executions to rule out random failure\n",
    " * Bad equipment:\n",
    "   * Cloning failed because of bad equipment, just try more executions\n",
    "   \n",
    "**Resource cost:** \n",
    "* **200 EUR each cloning experiment**\n",
    "\n",
    "**Input:**\n",
    "* **`mySequenceID`: small identifier for the sequence (string)**\n",
    "* **`myPromoter1`: promoter, 40nt from [ACGT] (string)**\n",
    "* **`myPrimer1`: primer, 15-30nt complementary to sequence (string)**\n",
    "* **`myTm`: melting Temperature, number (integer)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each cloning, the Clone_ID, the promoter sequence, the corresponding primer and the melting temperature must be given.\n",
    "\n",
    "# The following lines of code are an example of these variables.\n",
    "# To be able to clone further sequences, you have to define another set of all variables including the Clone_ID for each cloning.\n",
    "# To define a further Clone_ID you for example have to replace the number 1 in Clone_ID1 and Clone_1 with a different number.\n",
    "mySequenceID = None\n",
    "myPromoter = None\n",
    "myPrimer = None\n",
    "myTm = None\n",
    "\n",
    "newGene1 = CraftedGene( name=mySequenceID, prom=Seq(myPromoter), orf=\"GGGGGGGGGG\" )\n",
    "newHost1, clone_outcome1 = exp.clone_with_recombination( 'origin', Seq(myPrimer), gene=newGene1, tm=myTm )\n",
    "\n",
    "# mySequenceID = None\n",
    "# myPromoter = None\n",
    "# myPrimer =   None\n",
    "# myTm =  None\n",
    "\n",
    "# newGene2 = CraftedGene( name=mySequenceID, prom=Seq(myPromoter), orf=\"GGGGGGGGGG\" )\n",
    "# newHost2, clone_outcome2 = exp.clone_with_recombination( host, Seq(myPrimer), gene=newGene2, tm=myTm )\n",
    "\n",
    "print(\"Clone Outcome: \" + clone_outcome1)\n",
    "\n",
    "exp.print_status()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1.2 Measurement of the promoter strength\n",
    "The promoter strength represents expression per cell. Later, it is multiplied by the growth rate and the biomass concentration in order to determine the expression rate. In the following experiment the promoter strength is determined. \n",
    "\n",
    "**Resource cost:**\n",
    "* **100 EUR each experiment**\n",
    "\n",
    "**Input:**\n",
    "* **`newHost`**: hosts id generated in cloning experiment, e.g. 'origin.1'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "newHost = None\n",
    "promoter_act1 = exp.measure_promoter_strength( newHost, mySequenceID )\n",
    "promoter_data = combine_data([promoter_act1])\n",
    "promoter_data.display_data()\n",
    "exp.print_status()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1.3 Measurement of the final vaccine expression rate\n",
    "Now that you have tested some promoter sequences, perform the production experiments with the promoter sequence (Clone_ID) and use the optimal growth temperature, the maximum growth rate and the maximum biomass from the strain growth characterization task. \n",
    "\n",
    "Finally, the experimental results are exported to a csv file for further data analysis. The function works without further user input. The output file is stored in the `Data` subfolder as `Production_Simulation.csv`.\n",
    "\n",
    "**Resources:**\n",
    "* **500 EUR each experiment**\n",
    "\n",
    "**Input:**\n",
    "* **in `simulate_vaccince_production`: host variable name (string), gene variable name (string), opt. Temp (int), Opt. Growth rate (float), Opt. Biomass (int)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To perform the production experiment replace None with the Clone_ID variable name of your best performing clone, the optimal growth temperature, the corresponding maximum growth rate and the maximum biomass (in this order).\n",
    "vacProd1 = exp.simulate_vaccine_production(host_names=[None, None], gene_name=None, cult_temp=None, growth_rate=None, biomass=None)\n",
    "\n",
    "if vacProd1.error:\n",
    "    print('Experiment failed.')\n",
    "else:\n",
    "    filepath = os.path.join('..','Data','Production_Simulation.csv')\n",
    "    # add variable names for additional vaccine production experiment results here:\n",
    "    vacProdAll = combine_data([vacProd1]) # e.g. [vacProd1, vacProd2]\n",
    "    vacProdAll.export_data(filepath)\n",
    "    print('Clone Outcome production stored in: {}'.format(filepath))\n",
    "    \n",
    "exp.print_status()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.2 Data analysis of promoter strength\n",
    "\n",
    "The biotechnological goal is the construction of a host strain with high productivity. Moreover, scientifically we like to investigate the relationship between promoter strength and GC content. First, you will examine your own data in a plot for correlation, subsequently, all groups will enter their results for the species specific promoter strength versus GC content plot in an online plot. The online plot shows the results of all groups and allows a more solid conclusion.\n",
    "\n",
    "\n",
    "### 4.2.1 Visualization of the results\n",
    "\n",
    "Summarize your results from the laboratory workflow in a graph by plotting the promoter strengths of the final relative expression rates against the respective GC contents of the promoter sequences. You can perform the data analysis in Excel by importing the data file `Production_Simulation.csv` and generating a scatter plot of relative expression rate versus GC-content. Alternatively, you can use the Python code with scrambled lines (interactive web-site: [in this link](http://parsons.problemsolving.io/puzzle/7cd06ab7e8c94c7eba31bc50ac07a0a7)).\n",
    "\n",
    "Enter the correct code sequence in the cell below. The code cell that extracts the data columns for GC content and relative expression rate is missing the column number for the corresponding data. Add these column number in the fields labeled with `None` and note that Python starts counting from zero, thus the first column has index `0`.\n",
    "\n",
    "**Input:**\n",
    "* **`my_data`: GC-content column name (string)**\n",
    "* **`my_data`: expression rate column name (string)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load Snippets/rev_ExprPlot.py\n",
    "# Insert the correct code sequence for plotting in this cell.\n",
    "# %load Snippets/snip_ExprPlot.py \n",
    "\n",
    "my_data = pd.read_csv(DataFile)\n",
    "DataFile = os.path.join('Data','Production_Simulation.csv')\n",
    "GCcont, Express = my_data[None].values, my_data[None].values\n",
    "plt.gca().set(xlabel='GC-cont', ylabel='rel. expression')\n",
    "plt.savefig('RelExpress_Vs_GCcont_allProm.png', format='png')\n",
    "plt.plot(GCcont,Express, linestyle = '--', marker = 'x', color = 'grey')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Package dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -m -p IPython,ipywidgets,matplotlib,numpy,pandas,openpyxl,sklearn,scipy,joblib,watermark"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading of libraries if not automatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install pandas\n",
    "%pip install silvio\n",
    "%pip install scikit-learn"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bls_py39",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
